### PR TITLE
feat: add recovery.

### PR DESCRIPTION
**Pull request by gdethier**
logion-network/logion-internal#454

### CODE REVIEW

File: packages/client/test/Recovery.spec.ts [hunk 1]
```diff
@@ -0,0 +1,783 @@
...
+
+    it("builds an initial no protection state", async () => {
+        const data: FetchAllResult = {
+            pendingProtectionRequests: [],
+            acceptedProtectionRequests: [],
+            rejectedProtectionRequests: [],
+            recoveryConfig: undefined,
+            recoveredAddress: undefined,
+        };
+        testGetInitialState(data, NoProtection);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:29]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 2]
```diff
@@ -0,0 +1,783 @@
...
+                buildAcceptedAliceRequest(),
+                buildAcceptedBobRequest()
+            ],
+            rejectedProtectionRequests: [],
+            recoveryConfig: {
+                legalOfficers: legalOfficers.map(legalOfficer => legalOfficer.address)
+            },
+            recoveredAddress: undefined,
+        };
+        testGetInitialState(data, ActiveProtection);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:45]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 3]
```diff
@@ -0,0 +1,783 @@
...
+                buildAcceptedAliceRecoveryRequest(),
+                buildAcceptedBobRecoveryRequest(),
+            ],
+            rejectedProtectionRequests: [],
+            recoveryConfig: {
+                legalOfficers: legalOfficers.map(legalOfficer => legalOfficer.address)
+            },
+            recoveredAddress: RECOVERED_ADDRESS,
+        };
+        testGetInitialState(data, ClaimedRecovery);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:61]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 4]
```diff
@@ -0,0 +1,783 @@
...
+            pendingProtectionRequests: [],
+            acceptedProtectionRequests: [
+                buildAcceptedAliceRequest(),
+                buildAcceptedBobRequest()
+            ],
+            rejectedProtectionRequests: [],
+            recoveryConfig: undefined,
+            recoveredAddress: undefined,
+        };
+        testGetInitialState(data, AcceptedProtection);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:75]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+            pendingProtectionRequests: [
+                buildPendingAliceRecoveryRequest(),
+                buildPendingBobRecoveryRequest(),
+            ],
+            acceptedProtectionRequests: [],
+            rejectedProtectionRequests: [],
+            recoveryConfig: undefined,
+            recoveredAddress: undefined,
+        };
+        testGetInitialState(data, PendingProtection);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:89]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+                buildPendingBobRecoveryRequest()
+            ],
+            acceptedProtectionRequests: [
+                buildAcceptedAliceRecoveryRequest()
+            ],
+            rejectedProtectionRequests: [],
+            recoveryConfig: undefined,
+            recoveredAddress: undefined,
+        };
+        testGetInitialState(data, PendingProtection);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:104]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 7]
```diff
@@ -0,0 +1,783 @@
...
+            pendingProtectionRequests: [],
+            acceptedProtectionRequests: [
+                buildAcceptedAliceRecoveryRequest(),
+                buildAcceptedBobRecoveryRequest(),
+            ],
+            rejectedProtectionRequests: [],
+            recoveryConfig: undefined,
+            recoveredAddress: undefined,
+        };
+        testGetInitialState(data, PendingRecovery);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:118]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 8]
```diff
@@ -0,0 +1,783 @@
...
+                buildAcceptedAliceRecoveryRequest(),
+                buildAcceptedBobRecoveryRequest(),
+            ],
+            rejectedProtectionRequests: [],
+            recoveryConfig: {
+                legalOfficers: legalOfficers.map(legalOfficer => legalOfficer.address)
+            },
+            recoveredAddress: RECOVERED_ADDRESS,
+        };
+        testGetInitialState(data, ClaimedRecovery);
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:134]: Missing `await`.

File: packages/client/test/Recovery.spec.ts [hunk 9]
```diff
@@ -0,0 +1,783 @@
...
+
+                factory.setupDefaultNetworkState();
+                factory.setupAuthenticatedDirectoryClientMock(LOGION_CLIENT_CONFIG, token.value);
+
+                const nodeApi = factory.setupNodeApiMock(LOGION_CLIENT_CONFIG);
+                const emptyOption = {
+                    isEmpty: true,
+                    isNone: true,
+                    unwrap: () => ({})
+                };
```
Comment (benoitdevos) [inline packages/client/test/Recovery.spec.ts:475]: It worth defining `emptyOption` once for all tests using it.

**Review by benoitdevos — APPROVED** [review]
Great job !!
A couple of minor points to fix.

File: packages/client/src/Recovery.ts [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+        this.sharedState = sharedState;
+    }
+
+    private sharedState: SharedStateWithLegalOfficers;
+
+    async claimRecovery(signer: Signer): Promise<ClaimedRecovery> {
+        if(this.sharedState.recoveredAddress === undefined) {
+            throw new Error("No recovery was request");
+        }
+        signer.signAndSend({
```
[the remark this answers is not in the document]
Comment (gdethier) [reply packages/client/src/Recovery.ts:300]: Yes

**Pull request merged**